code/datums/actions/xeno_action.dm

```diff
@@ -112,33 +111,36 @@
 //checks if the linked ability is on some cooldown.
 //The action can still be activated by clicking the button
 /datum/action/xeno_action/proc/action_cooldown_check()
-	return !on_cooldown
+	return !cooldown_id
+
 
 /datum/action/xeno_action/proc/clear_cooldown()
-	for(var/timer in active_timers)
-		qdel(timer)
-		on_cooldown_finish()
+	if(!cooldown_id)
+		return
+	deltimer(cooldown_id)
+	on_cooldown_finish()
+
 
 /datum/action/xeno_action/proc/get_cooldown()
 	return cooldown_timer
 
+
 /datum/action/xeno_action/proc/add_cooldown()
-	if(!length(active_timers)) // stop doubling up
-		last_use = world.time
-		on_cooldown = TRUE
-		cooldown_id = addtimer(CALLBACK(src, .proc/on_cooldown_finish), get_cooldown(), TIMER_STOPPABLE)
-		button.overlays += cooldown_image
-		update_button_icon()
+	if(cooldown_id) // stop doubling up
+		return
+	last_use = world.time
+	cooldown_id = addtimer(CALLBACK(src, .proc/on_cooldown_finish), get_cooldown(), TIMER_STOPPABLE)
+	button.overlays += cooldown_image
+	update_button_icon()
+
 
 /datum/action/xeno_action/proc/cooldown_remaining()
-	for(var/i in active_timers)
-		var/datum/timedevent/timer = i
-		return (timer.timeToRun - world.time) * 0.1
-	return 0
+	return timeleft(cooldown_id) * 0.1
+
 
 //override this for cooldown completion.
 /datum/action/xeno_action/proc/on_cooldown_finish()
-	on_cooldown = FALSE
+	cooldown_id = null
```

Comment:
```suggestion
	qdel(cooldown_id)
```
to actually stop the timer properly

Reply:
This proc is called by `clear_cooldown()`, which stops the timer.

Reply:
May be safer to clear it there, though...

Reply:
then it should be done like that there, it's bad to split these two lines apart

Reply:
this suggestion is wrong since on_cooldown_finish is called when the timer ends

Reply:
This is coded correctly no need to change anything now.